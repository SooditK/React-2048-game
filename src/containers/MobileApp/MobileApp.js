import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Board from '../../components/Board';
import swipeDetect from '../../utils/mobileEvents';
import styles from './mobileApp.scss';
import Scores from '../../components/Scores';
import GameOverCard from '../GameOver';
import Speaker from '../../components/Speaker';
import WrapperButton from '../../components/WrapperButton';
import resetSvg from '../../assets/svg/reset.svg';
import undoSvg from '../../assets/svg/undo.svg';

// Mobile application entry
export default class MobileApp extends Component {
  static propTypes = {
    matrix: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)).isRequired,
    onMoveUp: PropTypes.func.isRequired,
    onMoveDown: PropTypes.func.isRequired,
    onMoveLeft: PropTypes.func.isRequired,
    onMoveRight: PropTypes.func.isRequired,
    onPlaceRandom: PropTypes.func.isRequired,
    onReset: PropTypes.func.isRequired,
    onUndo: PropTypes.func.isRequired,
    isMoved: PropTypes.bool.isRequired,
    audioMove: PropTypes.instanceOf(Audio).isRequired,
    audioPopup: PropTypes.instanceOf(Audio).isRequired,
    pastLen: PropTypes.number.isRequired,
    score: PropTypes.number,
    bestScore: PropTypes.number,
    gameOver: PropTypes.bool
  };

  static defaultProps = {
    score: 0,
    bestScore: 0,
    gameOver: false
  };

  constructor(...args) {
    super(...args);

    this.state = {
      speakerOn: true
    };
  }

  componentDidMount() {
    if (this.gameBoard) {
      swipeDetect(this.gameBoard, this.directionCallback);
    }
  }

  directionCallback = direction => {
    switch (direction) {
      case 'up':
        this.handleMoveUp();
        break;
      case 'down':
        this.handleMoveDown();
        break;
      case 'left':
        this.handleMoveLeft();
        break;
      case 'right':
        this.handleMoveRight();
        break;
      default:
        break;
    }
  };

  generalMove = func => {
    func();
    const { isMoved } = this.props;
    const { speakerOn } = this.state;
    if (speakerOn && isMoved) {
      this.props.audioMove.play();
    }
    setTimeout(() => {
      this.props.onPlaceRandom();
      if (speakerOn && isMoved) {
        this.props.audioPopup.play();
      }
    }, 100);
  };

  handleMoveUp = () => {
    this.generalMove(this.props.onMoveUp);
  };

  handleMoveDown = () => {
    this.generalMove(this.props.onMoveDown);
  };

  handleMoveLeft = () => {
    this.generalMove(this.props.onMoveLeft);
  };

  handleMoveRight = () => {
    this.generalMove(this.props.onMoveRight);
  };

  handleUndo = () => {
    if (this.props.pastLen > 3) {
      this.props.onUndo(-2);
    }
  };

  handleSpeakerClick = speakerOn => {
    this.setState({
      speakerOn
    });
  };

  render() {
    const { matrix, bestScore, score, gameOver, onReset } = this.props;
    return (
      <div className={styles.mobileApp}>
        <div className={styles.head}>
          <h1 className={styles.title}>2048</h1>
          <div>
            <Scores bestScore={bestScore} score={score} />
            <div className={styles.rowBtn}>
              <div className={styles.btn}>
                <Speaker onClick={this.handleSpeakerClick} />
              </div>
              <WrapperButton onClick={this.handleUndo}>
                <img src={undoSvg} alt="undo" />
              </WrapperButton>
              <WrapperButton onClick={this.props.onReset} cls={styles.last}>
                <img src={resetSvg} alt="reset" />
              </WrapperButton>
            </div>
          </div>
        </div>
        <div
          id="gameBoard"
          ref={div => {
            this.gameBoard = div;
          }}
          className={styles.board}
        >
          <Board matrix={matrix} />
        </div>
        <GameOverCard gameOver={gameOver} score={score} onReset={onReset} />
      </div>
    );
  }
}
