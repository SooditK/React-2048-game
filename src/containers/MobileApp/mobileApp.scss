.mobile-app {
  padding: 8vw 4vw 0;
  text-align: center;

  .head {
    display: flex;
    justify-content: space-around;

    .title {
      font-size: 3rem;
      // margin-top: -2.5vw;
      text-align: left;
    }
  }

  .row-btn {
    align-items: center;
    display: flex;
    flex-grow: 1;
    margin: 3vw 0 6vw;

    .btn {
      margin-right: 4vw;
      margin-top: 4vw;
    }

    img {
      margin-bottom: -1vw;
      // width: 2.1vw;
    }

    .last {
      margin-right: -1.5vw;
    }
  }

  .board {
    display: flex;
    justify-content: center;
  }
}
