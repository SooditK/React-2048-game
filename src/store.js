import { createStore, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import undoable from 'redux-undo';
import rootReducer from './reducers';
import rootSaga from './sagas';

const initHistory = JSON.parse(localStorage.getItem('state') || 'null');

const sagaMiddleware = createSagaMiddleware();
/* eslint-disable no-underscore-dangle */
const args = [
  undoable(rootReducer, {
    limit: 11, // set a limit for the history
    ignoreInitialState: true
  })
];
// Load localStorage data if available
if (initHistory) {
  args.push(initHistory);
}

if (window.__REDUX_DEVTOOLS_EXTENSION__) {
  args.push(
    compose(
      applyMiddleware(sagaMiddleware),
      // Redux devtools necessary code
      window.__REDUX_DEVTOOLS_EXTENSION__()
    )
  );
} else {
  args.push(applyMiddleware(sagaMiddleware));
}

const store = createStore(...args);
sagaMiddleware.run(rootSaga);

// Call this function while redux state changed,
// this callback save redux state to localStorage
store.subscribe(() => {
  const state = store.getState();
  localStorage.setItem('state', JSON.stringify(state));
});

export default store;
