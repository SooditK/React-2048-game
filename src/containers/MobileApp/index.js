import { connect } from 'react-redux';
import { ActionCreators } from 'redux-undo';
import {
  moveUp,
  moveDown,
  moveLeft,
  moveRight,
  placeRandom,
  reset
} from '../../reducers/board';
import MobileApp from './MobileApp';

const mapStateToProps = state => ({
  matrix: state.present.board.present.matrix,
  isMoved: state.present.board.present.isMoved,
  pastLen: state.past.length,
  score: state.present.board.present.score,
  bestScore: state.present.board.present.bestScore,
  gameOver: state.present.board.present.gameOver
});

const mapDispatchToProps = {
  onPlaceRandom: placeRandom,
  onMoveUp: moveUp,
  onMoveDown: moveDown,
  onMoveLeft: moveLeft,
  onMoveRight: moveRight,
  onReset: reset,
  onUndo: ActionCreators.jump // Undo move and generated cell
};

export default connect(mapStateToProps, mapDispatchToProps)(MobileApp);
