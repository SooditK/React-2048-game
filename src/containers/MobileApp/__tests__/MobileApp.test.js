import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import store from '../../../store';
import Container from '..';
import MobileApp from '../MobileApp';

Enzyme.configure({ adapter: new Adapter() });
const audio = new Audio();

describe('<MobileApp />', () => {
  it('component render', () => {
    const app = renderer
      .create(
        <Provider store={store}>
          <Container audioMove={audio} audioPopup={audio} />
        </Provider>
      )
      .toJSON();
    expect(app).toMatchSnapshot();
  });

  it('component instance', () => {
    const matrix = [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]];
    const moveUp = jest.fn();
    const moveDown = jest.fn();
    const moveLeft = jest.fn();
    const moveRight = jest.fn();
    const placeRandom = jest.fn();
    const reset = jest.fn();
    const undo = jest.fn();
    const mobileApp = mount(
      <MobileApp
        matrix={matrix}
        audioMove={audio}
        audioPopup={audio}
        onMoveUp={moveUp}
        onMoveDown={moveDown}
        onMoveLeft={moveLeft}
        onMoveRight={moveRight}
        onPlaceRandom={placeRandom}
        onReset={reset}
        onUndo={undo}
        isMoved
        pastLen={4}
      />
    );
    const ins = mobileApp.instance();
    ins.directionCallback();
    ins.directionCallback('up');
    jest.runAllTimers();
    expect(moveUp.mock.calls.length).toBe(1);
    expect(placeRandom.mock.calls.length).toBe(1);

    ins.directionCallback('down');
    jest.runAllTimers();
    expect(moveDown.mock.calls.length).toBe(1);
    expect(placeRandom.mock.calls.length).toBe(2);

    ins.directionCallback('left');
    jest.runAllTimers();
    expect(moveLeft.mock.calls.length).toBe(1);
    expect(placeRandom.mock.calls.length).toBe(3);

    ins.directionCallback('right');
    jest.runAllTimers();
    expect(moveRight.mock.calls.length).toBe(1);
    expect(placeRandom.mock.calls.length).toBe(4);

    ins.handleUndo();
    jest.runAllTimers();
    expect(undo.mock.calls.length).toBe(1);

    ins.handleSpeakerClick(false);
    expect(mobileApp.state('speakerOn')).toBe(false);

    ins.directionCallback('right');
    jest.runAllTimers();
    expect(moveRight.mock.calls.length).toBe(2);
    expect(placeRandom.mock.calls.length).toBe(5);

    mobileApp.setProps({
      pastLen: 2
    });
    ins.handleUndo();
    jest.runAllTimers();
    expect(undo.mock.calls.length).toBe(1);
  });
});
