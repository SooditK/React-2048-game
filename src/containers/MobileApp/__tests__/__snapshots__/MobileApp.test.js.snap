// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MobileApp /> component render 1`] = `
<div
  className="mobileApp"
>
  <div
    className="head"
  >
    <h1
      className="title"
    >
      2048
    </h1>
    <div>
      <div
        className="index"
      >
        <p
          className="score"
        >
          SCORE
          <em
            className="text"
          >
            0
          </em>
        </p>
        <p
          className="score best"
        >
          BEST
          <em
            className="text"
          >
            0
          </em>
        </p>
      </div>
      <div
        className="rowBtn"
      >
        <div
          className="btn"
        >
          <button
            className="btn-flat btn-md btn-default"
            onClick={[Function]}
          >
            <img
              alt="speaker"
              src="speaker-on.svg"
            />
          </button>
        </div>
        <div
          className="btn "
        >
          <button
            className="btn-flat btn-md btn-default"
            onClick={[Function]}
          >
            <img
              alt="undo"
              src="undo.svg"
            />
          </button>
        </div>
        <div
          className="btn last"
        >
          <button
            className="btn-flat btn-md btn-default"
            onClick={[Function]}
          >
            <img
              alt="reset"
              src="reset.svg"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="board"
    id="gameBoard"
  >
    <table
      className="board"
    >
      <tbody>
        <tr>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
        </tr>
        <tr>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
          <td>
            <div
              className="cell"
            >
              <div
                className="number"
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div>
    <div
      className="hidden"
    >
      <div
        className="overlay"
      />
      <div
        className="modal center"
      >
        <div>
          <div
            className="gameover"
          >
            <div
              className="text"
            >
               
              SCORE
               
            </div>
            <div
              className="banner"
            >
              <p>
                0
              </p>
            </div>
          </div>
          <div
            className="buttonWrapper"
          >
            <button
              className="button"
              onClick={[Function]}
            >
              <img
                alt="reset"
                src="reset.svg"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
